### DISCUSSION TITLE
CheckRulesAsync on a property

### DISCUSSION THREAD

**Discussion by mtavares628** [MarimerLLC/csla#3390] [Questions]
I have some code from CSLA 4 that I'm converting to CSLA 6/7 so that I can use my business objects with Blazor.  One of my BOs has a function on it that checks some business logic by calling BusinessRules.CheckRules on a specific property, and then looks to see if the rule name exists in the list of broken rules.  This worked fine in my old ASP.NET 4.x application where everything I did was synchronous.  However, now that I'm using blazor, a number of my rules are now asynchronous because they are hitting a database as part of the logic. 

I had assumed that I could then make my check rules method asynchronous and just call BusinessRules.CheckRulesAsync on the same property. However, it appears there is no overload of CheckRulesAsync that takes just a property.  Is this an oversight, or should I still be able to just call the sync CheckRules method, and all of my async rules will get called without an issue?

So my old code looked like this:

```c#
public bool AnswerIsAssociatedWithFund(IPropertyInfo prop, string RuleToCheck)
        {
            BusinessRules.CheckRules(prop);
            if (BrokenRulesCollection.Count > 0)
            {
                foreach (BrokenRule item in BrokenRulesCollection)
                {
                    if ((item.Property ?? "") == (prop.Name ?? "") && item.RuleName.ToLower().Contains(RuleToCheck.ToLower()))
                    {
                        return false;
                    }
                }
            }
            return true;
        }
```
I was thinking I needed something like this:

```c#
public async Task<bool> AnswerIsAssociatedWithFundAsync(IPropertyInfo prop, string RuleToCheck)
        {
            await BusinessRules.CheckRulesAsync(prop);
            if (BrokenRulesCollection.Count > 0)
            {
                foreach (BrokenRule item in BrokenRulesCollection)
                {
                    if ((item.Property ?? "") == (prop.Name ?? "") && item.RuleName.ToLower().Contains(RuleToCheck.ToLower()))
                    {
                        return false;
                    }
                }
            }
            return true;
        }
```

However, BusinessRules.CheckRulesAsync(prop) is not a valid overload.  So again, my question, will the original method still work, if I called it from the blazor client (I'm thinking not), or should there be a new overload for BusinessRules.CheckRulesAsync that will take a property?

**Comment by jhuerta3** [comment]
Hello @mtavares628 
Maybe this can help you. I used the code below for a Async rule on Blazor. 

    public class EquipmentAssignedRuleAsync : BusinessRuleAsync
    {
        public EquipmentAssignedRuleAsync(IPropertyInfo primaryProperty)
          : base(primaryProperty)
        {
            InputProperties.Add(PrimaryProperty);
        }

        protected override async Task ExecuteAsync(IRuleContext context)
        {
            var id = (Guid)context.InputPropertyValues[PrimaryProperty];
            var dp = context.ApplicationContext.GetRequiredService<IDataPortal<EquipmentAssignedChecker>>();
            var checker = await dp.FetchAsync(id);
            if (checker.IsAssigned)
                context.AddErrorResult("The equipment cannot be deleted since it has been assigned to an employee.");
        }
    }


On BusinessObject I called like so. 

        protected override void AddBusinessRules()
        {
            base.AddBusinessRules();
            BusinessRules.AddRule(new EquipmentAssignedRuleAsync(IdProperty));
        }

**Reply by mtavares628** [reply]
Hi @jhuerta3 , thanks for your response. However, I am not having an issue defining an async business rule, I actually have a bunch of them.  I'm looking to use a method to force the rules to run on a specific property, and then return whether the rules passed or not.  I just don't see an overload of CheckRulesAsync that will run on a single property like CheckRules does.  My question is whether there should be an overload for this, or if I should still be able to run the synchronous CheckRules method from my blazor code without an issue.

**Reply by rockfordlhotka** [reply]
Would you expect that `CheckRulesAsync(property)` to let you await the response, or just kick off the rule?

Without giving this deep thought, it seems like a reasonable idea to me.

**Reply by mtavares628** [reply]
I would like to be able to await the response.  My second code example above illustrates what I was thinking I would need to happen to make my existing method logic work.

But I'm also not 100% sure whether it is necessary or not, which was why I was asking.   So right now, when a property changes, BusinessRules.CheckRules(<Property>) is being called.  I'm assuming that this is able to run both the sync and async rules from the blazor client without issue.  So maybe I can keep calling the regular CheckRules method?  But I may be missing something in how you are running the rules, so I'm looking for clarity based on what I'm trying to do, and whether or not that CheckRulesAsync overload is actually necessary.

**Reply by rockfordlhotka** [reply]
A call to `CheckRules()` runs all rules.

The ability to do `CheckRules(property)` exists primarily because a change to a child might require that you handle the `ChildChanged` event and explicitly run rules for a property on an object higher in the object graph. That call would run all rules for the specified property (sync and async).

Because this is in an event handler, it wouldn't normally be a good idea to await that call. Though maybe there are cases?

In your example, if I may ask, why don't you have an `IsAccociatedWithFund` property that is set by a rule based on changes to whatever other properties or events might change that value?

**Reply by mtavares628** [reply]
So in my use case, a property change may not be the only way that something is no longer associated with a fund.  A database change to a lookup table made by an admin outside of the scope of my business object may trigger the selected value to no longer be associated.  so I have a method that runs the CheckRules on the property that hits the DB at a specific point in time to make the check.

**Comment by kcabral817** [comment]
Interesting question.
